cmd/ipfs/rotate.go

```diff
@@ -0,0 +1,91 @@
+package main
+
+import (
+	"io"
+	"os"
+
+	cmds "github.com/ipfs/go-ipfs-cmds"
+	config "github.com/ipfs/go-ipfs-config"
+	oldcmds "github.com/ipfs/go-ipfs/commands"
+	fsrepo "github.com/ipfs/go-ipfs/repo/fsrepo"
+	"github.com/ipfs/interface-go-ipfs-core/options"
+)
+
+var rotateCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Rotates the ipfs identity.",
+		ShortDescription: `
+Generates a new ipfs identity and saves it to the ipfs config file.
+The daemon must not be running when calling this command.
+
+ipfs uses a repository in the local file system. By default, the repo is
+located at ~/.ipfs. To change the repo location, set the $IPFS_PATH
+environment variable:
+
+    export IPFS_PATH=/path/to/ipfsrepo
+`,
+	},
+	Arguments: []cmds.Argument{},
+	Options: []cmds.Option{
+		cmds.IntOption(bitsOptionName, "b", "Number of bits to use in the generated RSA private key.").WithDefault(nBitsForKeypairDefault),
+	},
+	PreRun: func(req *cmds.Request, env cmds.Environment) error {
+		cctx := env.(*oldcmds.Context)
+		daemonLocked, err := fsrepo.LockedByOtherProcess(cctx.ConfigRoot)
+		if err != nil {
+			return err
+		}
+
+		log.Info("checking if daemon is running...")
+		if daemonLocked {
+			log.Debug("ipfs daemon is running")
+			e := "ipfs daemon is running. please stop it to run this command"
+			return cmds.ClientError(e)
+		}
+
+		return nil
+	},
+	Run: func(req *cmds.Request, res cmds.ResponseEmitter, env cmds.Environment) error {
+		cctx := env.(*oldcmds.Context)
+		nBitsForKeypair, _ := req.Options[bitsOptionName].(int)
+		return doRotate(os.Stdout, cctx.ConfigRoot, nBitsForKeypair)
+	},
+}
+
+func doRotate(out io.Writer, repoRoot string, nBitsForKeypair int) error {
+	// Open repo
+	if err := checkWritable(repoRoot); err != nil {
+		return err
+	}
+	repo, err := fsrepo.Open(repoRoot)
+	if err != nil {
+		return err
+	}
+	defer repo.Close()
+
+	// Read config file from repo
+	cfg, err := repo.Config()
+	if err != nil {
+		return err
+	}
+
+	// Save old identity to keystore
+	privKey, err := cfg.Identity.DecodePrivateKey("") //XXX
+	if err != nil {
+		return err
+	}
+	keystore := repo.Keystore()
+	if err := keystore.Put(cfg.Identity.PeerID, privKey); err != nil {
+		return err
+	}
```

Comment:
What do you think about using some format like `prevKey-{Timestamp}` and appending `-{conflictNumber}` if the key name already exists? Might be easier for people to track/handle.

Additionally (or as a replacement) we should have a flag that allows specifying the name of the output key (and returning an error if the name already exists).

Reply:
What is the conflict number for? There is an alternative way to make the name unique, e.g.
KeyRotation-PeerID-FromPublicKey-ToPublicKey-Timestamp
(The time stamp is just for convenience, not necessary for uniqueness.)

Reply:
Take a look at latest commit. Specifically:
* I have to places marked as XXX, where I am not sure what "passphrase" to pass DecodePrivateKey.
* See how you like the key name generation

Reply:
>I have to places marked as XXX, where I am not sure what "passphrase" to pass DecodePrivateKey.

Don't worry about it. If you look at the function you'll see we don't even use the passphrase, it's just a TODO.

> See how you like the key name generation

`From -> To` seems like overkill. At that point you might as well just do `From` since that contains all the unique information about the key. I'm a little worried about not knowing what base representation people expect to see their keys in. WDYT about circumventing this problem entirely for the time being and just forcing people to pass in the name to use for their old key (and checking to see if it exists so we don't overwrite it). This might let us ignore this problem until we figure out our story on how we're going to represent PeerIDs of various types throughout IPFS (e.g. do we use base36 everywhere?)